lang/lang/statics/base.stx

```diff
@@ -0,0 +1,359 @@
+  orB(a, b) = FALSE() :- false | error $[Bug in Statix spec: no matching rule for orB([a], [b]).].
+
+  notB : BOOLEAN -> BOOLEAN
+  notB(TRUE()) = FALSE().
+  notB(FALSE()) = TRUE().
+
+rules // Declararing and resolving
+
+  // For declaring and resolving modules in relation `mod` and `mod_wip`, see
+  // statics/module!declareModule and statics/module!resolveModuleList
+
+  /**
+   * Declare a value [name] with type [TYPE] and ValKind [kind] in scope [s],
+   * set the type on the value name.
+   */
+  declareVal : scope * string * TYPE * ValKind
+  declareVal(s, name, type, kind) :-
+    declareValWithoutSettingType(s, name, type, kind),
+    @name.type := type.
+
+  /**
+   * Declare a value [name] with type [TYPE] and ValKind [kind] in scope [s],
+   * do not set the type on the value name.
+   */
+  declareValWithoutSettingType : scope * string * TYPE * ValKind
+  declareValWithoutSettingType(s, name, type, kind) :-
+    !val[name, type] in s,
+    @name.kind := kind,
+    resolveVal(s, name) == [_]
+      | error $[Duplicate definition of value [name]].
+
+  /**
+   * Resolve a value [name] in scope [s]
+   */
+  resolveVal : scope * string -> list((path * (string * TYPE)))
+  resolveVal(s, name) = occs :-
+    query val
+      filter P*
+         and { name' :- name' == name }
+         min and false
+          in s |-> occs.
+
+  /**
+   * resolveUniqueVal(s, name) = (ty, resolved)
+   * Resolves a value [name] to a unique type [ty] in scope [s].
+   * Sets a type and a reference to the declaration on this name.
+   * Gives a undeclared value error if no declaration is found.
+   * Does not give an error if there are multiple definitions as there are
+   * already errors on the declarations.
+   * [resolved] indicated whether [name] resolved to a single declaration.
+   * If it did not (resolved == FALSE()), [ty] will be an unbound variable.
+   * See also resolveVal
+   */
+  resolveUniqueVal : scope * string -> (TYPE * BOOLEAN)
+  resolveUniqueVal(s, name) = resolveUniqueVal_1(resolveVal(s, name), name).
+
+    resolveUniqueVal_1 : list((path * (VALID * TYPE))) * VALID -> (TYPE * BOOLEAN)
+    resolveUniqueVal_1([], name) = (_, FALSE()) :-
+      false | error $[Cannot resolve [name]] @name.
+
+    resolveUniqueVal_1([(_, (name', ty))], name) = (ty, TRUE()) :-
+      @name.ref := name',
+      @name.type := ty.
+
+    // If there are multiple, the declaration will give an error.
+    resolveUniqueVal_1([_,_|_], _) = (_, FALSE()).
+
+
+  /**
+   * Declare a function [name] with type [TYPE] in scope [s], set the type on
+   * the function name.
+   */
+  declareFunc : scope * string * TYPE * FuncImplKind
+  declareFunc(s, name, type, kind) :-
+    !func[name, type] in s,
+    resolveFunc(s, name) == [_] | error $[Duplicate function [name]] @name,
+    @name.type := type,
+    @name.kind := kind,
+    @name.qualifiedModule := getQualifiedModule(getFileIfExists(s), s).
+
+    getQualifiedModule : list(scope) * scope -> list(MODID)
```

Comment:
In many places constraint rules seem to have additional indentation. Is this indentation intended? If not, can you fix this and other instances?

Reply:
This is intended. If a rule has more indentation it is more or a less a helper method for the rule above. This means that it is not used in other rules.
(There are some explicit helper methods, e.g. `resolveUniqueFunc` with `resolveUniqueFunc_1`, but I tried to give them more descriptive names if possible)